### ISSUE TITLE
How to compile TEdit with Visual Studio 2017

### ISSUE THREAD

**Issue by ckeyboard** [TEdit/Terraria-Map-Editor#1070]
Hello,
I'm currently I'm trying to compile TEdit through Visual Studio 2017, and have run into compiler issues that are preventing me from creating a successful build. Before attempting the .bat build script I've accomplished the following:
1. installed the latest plugins for Visual Studio 2017
2. executed nuget.exe at least once to make sure MSBuild is up-to-date
3. executed the TEditNoInstaller.sln script through Visual Studio 2017
(Also, I'm running Windows 8.1, if it makes any difference.)

When executing the build script (.bat), I run into the following compiler errors:

1ST PROBLEM:
Project file contains ToolsVersion="12.0". This toolset may be unknown or missi
ng, in which case you may be able to resolve this by installing the appropriate
 version of MSBuild, or the build may have been forced to a particular ToolsVer
sion for policy reasons. Treating the project as if it had ToolsVersion="4.0".
For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.

2ND PROBLEM:
error MSB4175: The task factory "CodeTaskFactory"
 could not be loaded from the assembly "C:\Windows\Microsoft.NET\Framework\v4.0
.30319\Microsoft.Build.Tasks.v12.0.dll". Could not load file or assembly 'file:
///C:\Windows\Microsoft.NET\Framework\v4.0.30319\Microsoft.Build.Tasks.v12.0.dl
l' or one of its dependencies. The system cannot find the file specified.

The first problem I've put to the side, since it seems to me to be more of a warning message. As for the 2nd error, there is no "Microsoft.Build.Tasks.v12.0.dll" to be found within "C:\Windows\Microsoft.NET\Framework\v4.0.30319\", but there IS a "Microsoft.Build.Tasks.v4.0.dll" in that same directory. Unfortunately, if one is to change change assembly file from v12.0 to v4.0 (which I did), the compiler will then complain about the '$' symbol used throughout the source code (which I assume is string interpolation or some newer feature in csharp.)

So, my question is, how does one fix this issue? Does the assembly file need to point to some other file? Does TEdit need to be updated to work with Visual Studio 2017? Does Visual Studio need some setting tweaked to work with TEdit? Or is there an intermediate step missing?

Sincerely,
ckeyboard

(Side note: If I can get this working, I wouldn't mind submitting what I learned as a beginner tutorial, since I was unable to find more information.)

**Comment by BinaryConstruct** [comment]
I think this has to do with porting this project from VS2012. I've tried to switch all the dependencies to nuget packages. Let me know if this helps.

**Comment by ckeyboard** [comment]
So I have good news and bad news. 
The good news is that updating nuget.exe has allowed Visual Studio 2017 to pull in a number of missing dependencies. More than 20 packages were successfully installed upon executing the build script. 
The bad news is that basically all the problems mentioned from before are still showing up. This is the updated output from the console:

1ST PROBLEM:
Project file contains ToolsVersion="15.0". This toolset may be unknown or missi
ng, in which case you may be able to resolve this by installing the appropriate
 version of MSBuild, or the build may have been forced to a particular ToolsVer
sion for policy reasons. Treating the project as if it had ToolsVersion="4.0".
For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.

2ND PROBLEM:
C:\Users\...\Software\Terraria-Map-Editor-master\Terraria-Map-Editor-master\TEdi
tXna\TEditXna.csproj(582,5): error MSB4175: The task factory "CodeTaskFactory"
could not be loaded from the assembly "C:\Windows\Microsoft.NET\Framework\v4.0.
30319\Microsoft.Build.Tasks.Core.dll". Could not load file or assembly 'file://
/C:\Windows\Microsoft.NET\Framework\v4.0.30319\Microsoft.Build.Tasks.Core.dll'
or one of its dependencies. The system cannot find the file specified.

Just to rephrase what the error messages are saying, there is no "Microsoft.Build.Tasks.Core.dll" file. However, if one makes the project to point to "Microsoft.Build.Tasks.v4.0.dll" instead, the compiler is able to begin, but cannot understand what the '$' symbol means in the csharp code.

To be clear, these are the Microsoft.Build files that can be found inside the "C:\Windows\Microsoft.NET\Framework\v4.0.30319\" directory:
Microsoft.Build.Conversion.v4.0.dll
Microsoft.Build.dll
Microsoft.Build.Engine.dll
Microsoft.Build.Framework.dll
Microsoft.Build.Tasks.v4.0.dll
Microsoft.Build.Utilities.v4.0.dll
Microsoft.Build.xsd

This directory also includes:
MSBuild
MSBuild.exe
msbuild.exe.config

**Comment by ckeyboard** [comment]
I have great news! I successfully built TEdit by making changes to the source code. While there were at least 2 warning messages during the compilation phase, the program runs smoothly so far.
Here are the list of changes I made to the code:
1. Inside the "DateStampVersion.tasks" file, the "AssemblyFile" parameter needs to be adjusted to the correct path as follows:
AssemblyFile="C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Build.Tasks.Core.dll"
The reason for this change is that $(MSBuildToolsPath) will expand to "C:\Windows\Microsoft.NET\Framework\v4.0.30319\", which is the wrong directory.
Now there is probably a way to make $(MSBuildToolsPath) point to "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\" rather than using the absolute path, but I'll let the professionals figure that out. :-)
2. Inside the "build_beta.bat" build script, the 'msbuild' path variable (found on line 3) needs to be changed to the following string value:
set msbuild="C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\MSbuild.exe"
The reason for this change is that we want to use the up-to-date MSBuild executable that comes bundled with Visual Studio 2017, rather than the old one found in the .NET folder.

After making these 2 changes and running "build_beta.bat", the compiler will finish building the program. As a reminder, there are 2 warning messages which may affect the stability of TEdit; and if you would like me to post the output, I'll be happy to do so.
